### ISSUE TITLE
Browser push notifications

### ISSUE THREAD

**Issue by benruijl** [syncthing/syncthing#5329]
It would be nice if the web front-end of syncthing offers browser push notifications, similar to how it works for websites such as Youtube and Facebook.

The push notifications about file changes are the only reason I use syncthing-gtk.

**Comment by klausenbusk** [comment]
https://medium.com/izettle-engineering/beginners-guide-to-web-push-notifications-using-service-workers-cb3474a17679 shows how this can be done.

**Comment by benruijl** [comment]
Not necessarily, since it makes sense to just give syncthing the permission to do push notifications  in general. Then on the web frontend the user can have a settings page with a set of tick boxes to filter messages of a certain type.

The first version could simply be a message about a synchronization update.

**Comment by AudriusButkevicius** [comment]
Sure, that's still work that needs to be done. It's not enough just to add support for push notifications. Also, they only work while the tab is open I think?

**Comment by benruijl** [comment]
For reference, the GTK app has only 3 options:

- file updates
- synchronization completion
- errors

For the first version, we could simply have all 3 enabled.

**Comment by AudriusButkevicius** [comment]
Well if you want to contribute the feature that's great, but it's not that simple. If I throw in a folder with 1000s of tiny files l would not want to see 1000s of notifications, so this has to be thought through.

**Comment by benruijl** [comment]
I can help out, but it could take me quite some time to find my way in the project. 

It seems we need a javascript background task that can poll (or is able to receive) information about recent changes from Syncthing. Does a framework for this exist?

**Comment by klausenbusk** [comment]
> It seems we need a javascript background task that can poll (or is able to receive) information about recent changes from Syncthing. Does a framework for this exist?

javascript [fetch](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API) + https://docs.syncthing.net/dev/events.html + [Notifications API](https://developer.mozilla.org/en-US/docs/Web/API/Notifications_API). For a start you can just implement it the same way as syncthing-gtk (https://github.com/syncthing/syncthing-gtk/blob/master/syncthing_gtk/notifications.py).

**Comment by benruijl** [comment]
Thanks for the resources! This should be enough for a small mock-up :smiley:

**Comment by klausenbusk** [comment]
> Thanks for the resources! This should be enough for a small mock-up

If we want this to work without the tab opened, it gets a bit more complicated.

First we need a [Service Worker](https://developer.mozilla.org/en-US/docs/Web/API/Service_Worker_API) so we can use the [Push API](https://developer.mozilla.org/en-US/docs/Web/API/Push_API). [PushManager.subscribe()](https://developer.mozilla.org/en-US/docs/Web/API/PushManager/subscribe) gives us the endpoint, a public key and a "authentication secret". Syncthing need to store this information. With all this information on hand, Syncthing can now send notifications to the browser even when the tab is closed by using webpush (ex by using this Go package: https://github.com/SherClockHolmes/webpush-go).

**Comment by benruijl** [comment]
> If we want this to work without the tab opened, it gets a bit more complicated.

Since we want this, it seems the following modifications need to happen:
- Each Syncthing instance should generate a private and pubkey pair
- The web UI should have a small javascript snippet for registering a service worker (ie, this will generate the pop-up to subscribe). It will contain the pubkey of the instance.
- The result of the subscription should be sent back to syncthing via the web UI
- The notifications and what to send can now completely be handled in Syncthing (so in Go). There is no need for Syncthing's external event API. Internally it should listen to the events, construct notifications, and then broadcast them using webpush

Alternatively, we don't handle the notifications in Syncthing but use javascript with Syncthing's event API, but I don't know what the benefit would be.

Any thoughts?

**Comment by klausenbusk** [comment]
> Alternatively, we don't handle the notifications in Syncthing but use javascript with Syncthing's event API, but I don't know what the benefit would be.

That wouldn't work when the tab is closed.

**Comment by benruijl** [comment]
I'm not so sure about that. The Service Worker is run in the background by the browser. So if we code the logic for notifications there using the Syncthing event API, it will run even when the tab is closed.

**Comment by klausenbusk** [comment]
> I'm not so sure about that. The Service Worker is run in the background by the browser. So if we code the logic for notifications there using the Syncthing event API, it will run even when the tab is closed.

They are stopped after some time.

**Comment by tuarrep** [comment]
Any news on that?

I could draft a PR during my Christmas leaves to push browser notifications based on @Martchus list which sounds me a good starting point.

**Comment by benruijl** [comment]
That would be amazing!

**Comment by tuarrep** [comment]
@benruijl Have looked at this PR https://github.com/syncthing/syncthing/pull/6244? I need some feedbacks/testing on my implementation.

**Comment by benruijl** [comment]
I have not, but I will soon!

**Comment by agardnerIT** [comment]
I came across this as I'm interested in a related area.

Since this thread was created, newer frameworks and standards have become available. I wonder if we could leverage [OpenTelemetry](https://opentelemetry.io) so that each operation generates one or more spans.

The advantage to this method is:

- OpenTelemetry is an open standard, widely adopted, stable and understood
- Spans can be correlated with log files and other entities (such as hosts). So my backend system can say "show me hosts where failed file syncs occurred.
- The backend "span storage" solutions are widespread, mature and built to ingest the kind of scale data even if there were tens of thousands of files.

I also wonder if we could start with just "failed sync" events? As a user, that's probably what I care about most - when something goes wrong.
